File: src/main/bin/clean_up_disk_log.py
# -*- encoding=utf-8 -*-
"""
copy from CleanupDiskLog.pl
"""
#  Copyright (c) 2022. PengYunNetWork
#
#  This program is free software: you can use, redistribute, and/or modify it
#  under the terms of the GNU Affero General Public License, version 3 or later ("AGPL"),
#  as published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
#   without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
#   You should have received a copy of the GNU Affero General Public License along with
#   this program. If not, see <http://www.gnu.org/licenses/>.

import os
import sys


def find_project_folder(script_folder):
  """
  find project base folder from real script folder
  :param script_folder: like /var/testing/_packages/datanode-2.4.0/bin
  :return: like /var/testing/packages/datanode-2.4.0
  """
  path = script_folder

  # first find packages folder
  while "packages" not in os.listdir(path):
    parent = os.path.dirname(path)

    if parent == path:
      print("can't find datanode project folder, clean failed")
      exit(-1)
    else:
      path = parent

  path = os.path.join(path, "packages")

  # find datanode project folder
  for name in os.listdir(path):
    if "datanode" in name:
      return os.path.join(path, name)

  print("can't find datanode project folder, clean failed")
  exit(-1)


# like /var/testing/_packages/datanode-2.4.0/bin
script_folder = sys.path[0]
# like /var/testing/packages/datanode
project_folder = find_project_folder(script_folder)
disk_log_folder = os.path.join(project_folder, "var/storage/disklog")

if not os.path.exists(disk_log_folder):
  print("there is no disk log folder, don't need clean")
  exit(0)

# delete the oldest file, sort file by file name, delete the smallest
# only allow 3 file in folder
max_file_count = 3
filenames = os.listdir(disk_log_folder)
while len(filenames) > max_file_count:
  oldest_file = min(filenames)
  filenames.remove(oldest_file)
  os.remove(os.path.join(disk_log_folder, oldest_file))
  print("delete disk log:[%s]" % oldest_file)

File: src/main/bin/perl/CleanupDiskLog.pl
#!/usr/bin/perl

use warnings;
use strict;
use FindBin '$RealBin';
use File::Basename;

my $env_root = dirname($RealBin);
$env_root = &get_datanode_var_parent($env_root);
my $disk_log_dir = "$env_root/var/storage/disklog";

# get all files in disk log directory
opendir(TEMP, $disk_log_dir) || die "cannot open $disk_log_dir:$!";
my @files = readdir TEMP;
@files = grep { $_ ne "." and $_ ne ".." } @files;

my $max_files_count = 3;
my $total_count = scalar(@files);

if ($total_count <= $max_files_count) {
    die "current file count is $total_count\n";
}

# sorted the file list by file name
my @sorted_files = sort { $b cmp $a } @files;

for (my $index = $total_count -1; $index >= $max_files_count; $index--) {
    my $temp = "$disk_log_dir/$sorted_files[$index]";
    system("rm $temp");
}

sub getAllFilenamesFromFolder {
    my ($folder) = @_;

    my @filenames = ();
    if (!opendir(FOLDER, $folder)) {
        printf("Can't open dir:[$folder].\n");
        return @filenames;
    }

    foreach my $filename (readdir(FOLDER)) {
        if ($filename eq "." || $filename eq "..") {
            next;
        }

        printf("find file:[$filename] in folder:[$folder]\n");
        push(@filenames, $filename);
    }
    closedir(FOLDER);

    return @filenames;
}

# because the directory is link symblic, and we can not get the script path by `pwd`
# so we do like this
sub get_datanode_var_parent() {
    my ($old_base) = @_;    # like /var/testing/_packages/datanode-2.4.0
    my $packages_path = dirname(dirname($old_base))."/packages";    # like /var/testing/packages

    my @filenames = getAllFilenamesFromFolder($packages_path);

    foreach my $name (@filenames) {
        if ($name =~ /datanode/) {
            my $project_path = "$packages_path/$name"; # like /var/testing/packages/datanode
            printf("get datanode project folder:[$project_path]\n");
            return $project_path;
        }
    }

    die "can't find datanode project folder, exit\n";
    return "";
}

